### ISSUE TITLE
BUG: Issue with Pandas df.str.split with expand

### ISSUE THREAD

**Issue by phizzuela** [pandas-dev/pandas#35807]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python

>>> df=pd.DataFrame({'a':['Kickoff_2'],'b':['bla'],'c':[1],'x':[None],'y':[None],'z':[None]})
>>> df[['x','y','z']] = df['a'].str.split('_',expand=True,n=2)
...
    raise ValueError("Columns must be same length as key")
ValueError: Columns must be same length as key

>>> df_long=pd.DataFrame({'a':['Kickoff_2','Kickoff','Kickoff_pass_jump'],'b':['bla','bla','bla'],'c':[1,2,3],'x':[None,None,None],'y':[None,None,None],'z':[None,None,None]})
>>> df_long[['x','y','z']] = df_long['a'].str.split('_',expand=True,n=2)
>>> df_long
                   a    b  c        x     y     z
0          Kickoff_2  bla  1  Kickoff     2  None
1            Kickoff  bla  2  Kickoff  None  None
2  Kickoff_pass_jump  bla  3  Kickoff  pass  jump

```

#### Problem description

String + split() with expand only works if there is a row in the df that will later contain the desired number of columns. It would be nice if the first example here would get filled up with NaNs.

**Comment by dsaxton** [comment]
I don't think this is an issue with str.split, the error happens when attempting to assign to a DataFrame with something that has the wrong shape, which _should_ produce an error.

**Comment by rhshadrach** [comment]
Agreed @dsaxton. Does something like this help your use case @phizzuela?

````
names = ['x', 'y', 'z']
exploded = (
    df['a'].str.split('_',expand=True,n=len(names)-1)
    .rename(columns={k: name for k, name in enumerate(names)})
)
result = df.join(exploded)
````

This will be able to handle if there are fewer splits than `len(names)-1`.

**Comment by phizzuela** [comment]
Using the expand and n arguments in the pd.str.split() sounds like it should always return something of length n+1. Also it is pretty confusing that in this example the split only works when the last row of the example df is present. If expand together with n, would fill up with NaN values, everything would work, independent what what follows in the df.

**Comment by rhshadrach** [comment]
Ah - I see. I had previously missed the `expand` keyword. Thanks for pointing that out. The way I read the docs:

> If for a certain row the number of found splits < n, append None for padding up to n if expand=True

it seems to me your expectation is what should happen. That makes this a bug: when `n=2` and `expand=True`, the shape returned should _always_ have 3 columns (one plus the number of splits).

**Comment by AndreZoro** [comment]
Call it an expansion inspired by the incorrect documentation or a bug. It would be super nice to have the possibility to enforce the number of return values. In _pandas/core/strings.py_ you can find this note:
`
    Notes
    -----
    The handling of the 'n' keyword depends on the number of found splits:

    - If found splits > 'n',  make first 'n' splits only
    - If found splits <= 'n', make all splits
    - If for a certain row the number of found splits < 'n',
      append 'None' for padding up to 'n' if "expand=True"

    If using "expand=True", Series and Index callers return DataFrame and
    MultiIndex objects, respectively.
`
Looks like this was planned all along.

**Comment by simonjayhawkins** [comment]
> It would be super nice to have the possibility to enforce the number of return values.

or maybe another DataFrame method that allows padding rows and columns something similar to np.pad https://numpy.org/doc/stable/reference/generated/numpy.pad.html 

this would have more utility, and avoid a breaking API change on df.str.split(expand=True, n=...)

There is already an `DataFrame.pad` method whose semantics are different enough from NumPy that the `DataFrame.pad` should proably be deprecated. It is only an alias to `DataFrame.fillna(method='ffill')` after all.

Another future possibility is that using \_\_array_function__ we may be able pad a DataFrame to a prescribed number of columns in such a manner using the NumPy api.

of course, this doesn't address the return value of df.str.split directly, but the same would be achieved with an extra step.

**Comment by AndreZoro** [comment]
The thing about the example of this bug is that if you want to split the strings in a column at an underscore and expand it into there columns, it will only work depending on if there is one row where you have a string that contains at least two underscores. So if you have a string something like "part01_part02_part03" in at least one row, Pandas can expand the column into three new ones, if there is not a single row with these two underscores but with less, it does not work. Why this behavior can be interpreted as a documentation issue is a bit difficult to understand. I have tried fixing the code, but what Pandas does with the keyword "expand" is even harder to understand.

**Comment by rhshadrach** [comment]
@AndreZoro: I think one could argue that by padding the result to always contain 3 columns, we would be no longer be consistent with python's `str.split`. Is there a downside to having to do something like:

     df['a'].str.split('_', expand=True, n=2).pad(3)

?

**Comment by AndreZoro** [comment]
Well set _expand = False_ and stay consistent with the Python split.

And as long as there is one mere row which contains enough desired separators, it also expands with NaNs, which acording to your argument is then already inconsistent.

**Comment by AndreZoro** [comment]
Just tried your suggested workaround. It results in an error (using Pandas 1.1.1 on the computer I am at right now):
`TypeError: ffill() got an unexpected keyword argument "columns"`

I think for now, the cleanest workaround would be to append a row to the df which contains a string with enough separators in the hot column and then delete that row later again. I will try that out tomorrow.

**Comment by jouvin** [comment]
I was about to submit the very same issue after struggling with this problem. As explained by @AndreZoro, `split(expand=True)` and a maximum number of split works exactly as expected (padding with `np.NaN` or `None` the missing values) **except** if no row can be successfully splitted. This looks really as an inconsistent behaviour.

**Comment by rhshadrach** [comment]
> Just tried your suggested workaround. It results in an error (using Pandas 1.1.1 on the computer I am at right now): `TypeError: ffill() got an unexpected keyword argument "columns"`

This wasn't a workaround, it was proposed future behavior building off of https://github.com/pandas-dev/pandas/issues/35807#issuecomment-680961199. It seems to me this would (a) allow us not to change the behavior or increase the complexity of `str.split` and (b) provide a potentially useful method for other situations all while making the desired behavior here easy to accomplish.

**Comment by asishm** [comment]
I've run into this in the past few months where I needed the "pad" - you can do this via 

```py
In [6]: df['a'].str.split('_',expand=True,n=2).reindex(range(3), axis=1)
Out[6]:
         0  1   2
0  Kickoff  2 NaN
```
as answered here - https://stackoverflow.com/questions/69563497/pandas-string-split-into-multiple-columns-with-variable-number-of-delimited-va

**Comment by jouvin** [comment]
@asishm thanks for the link, it looks it is what I needed. It would be good that the `split()` documentation mention the need to reindex columns in case the split results in less columns than expected.

**Comment by rhshadrach** [comment]
+1 on adding to the documentation. I believe the `fillna` with None vs np.nan has an issue, I took a quick look but couldn't find it. Also regarding "pad", I'm finding pandas uses it as an alias for "fillna" pretty widely. I personally think of it as a "fill" whereas "pad" would alter the frame/series shape, but changing the meaning of pad as suggested in https://github.com/pandas-dev/pandas/issues/35807#issuecomment-680961199 (which I do like) seems to me to not be worth it.
